Distribute NFT tokens

zh opened the pull request:
Allow the faucet to **distribute NFT tokens** too. See _README_ file for more details about this kind of setup - `.env` settings etc.

jcramer commented:
I tested this code and ran into some issues.  

Seems like `NFT=no` functions fine, but `NFT=yes` keeps having problems with failing transactions due to an error `BCH amount too low`.  I wasn't able to get past the distribution phase for the NFT Group.

zh commented:
Sorry seems there is some confusion with PR... :( This is not my last version of the code. Will resubmit it again

zh commented:
Think I found the problem - [simpleTokenSend()](https://github.com/simpleledger/slp-faucet/pull/2/files#diff-87153a761a69dc3bdbe9cdeff43a7eb71aa7978b9a3874283b40658dbcf174d5R120), when sending burning-ready utxos was returning the remaining BCH to the main address '/0, so all subaddresses was left without any BCH. Moved NFT distribution **before** BCH one to prevent this.

Also addressed _502 error_ message problem by reducing number of call to the BCHD node - finding UTXO for burning executed one call per address, resulting in 19 calls. Reduced to 1 call getting all UTXOs for all addresses.

jcramer commented:
Ok, it looks like that worked.

For some reason when I try to include a documentHash for the NFT I get `Error: documentHash must be either 0 or 32 hex bytes`

jcramer commented on slpfaucet.ts:
```diff
@@ -217,7 +234,6 @@ export class SlpFaucetHandler {
     }
 
     public async nftChildTokenSend(groupId: string, inputUtxos: slpjs.SlpAddressUtxoResult[], tokenReceiverAddress: string, changeReceiverAddress: string): Promise<string> {
```
on line 240, it should be `const documentHash: Buffer|null = Buffer.from(process.env.DOCUMENTHASH!, "hex");`.   "hex" is missing.

zh replied:
fixed in commit# 54a464f

jcramer commented on server.ts:
```diff
@@ -19,29 +17,41 @@ app.use(bodyParser.urlencoded({ extended: true }));
-
         try {
-            await slpFaucet.evenlyDistributeTokens(process.env.TOKENID!);
+            if (process.env.NFT === 'yes') {
```
Lets keep this part of the server logic the same as master.  Token type logic should be handled in slpfaucet.ts.

jcramer commented on server.ts:
```diff
@@ -68,15 +78,28 @@ app.post("/", async (req, res) => {
     try {
         let inputs: slpjs.SlpAddressUtxoResult[] = [];
-        inputs = inputs.concat(changeAddr.balance.slpTokenUtxos[process.env.TOKENID!]).concat(changeAddr.balance.nonSlpUtxos);
+        if (process.env.NFT === 'yes') {
```
Same here, handle nft logic in slpfaucet.  Create a new method in slpfaucet to pick which send method to call.  This will keep the server.ts as clean as possible.

jcramer commented on slpfaucet.ts:
```diff
@@ -4,8 +4,8 @@ import { GrpcClient } from "grpc-bchrpc-node";
 
-const bitbox = new BITBOX();
-const client = new GrpcClient({url: "bchd.ny1.simpleledger.io" });
+const bitbox = new BITBOX({restURL: 'https://bchd.fountainhead.cash'});
```
We aren't using any of the network methods in BITBOX sdk, so we don't meed to use `restURL` parameter.

jcramer commented on slpfaucet.ts:
```diff
@@ -4,8 +4,8 @@ import { GrpcClient } from "grpc-bchrpc-node";
-const bitbox = new BITBOX();
-const client = new GrpcClient({url: "bchd.ny1.simpleledger.io" });
+const bitbox = new BITBOX({restURL: 'https://bchd.fountainhead.cash'});
+const client = new GrpcClient({url: 'bchd.fountainhead.cash' });
```
Not specific to NFTs, but lets go ahead use a new environment variable `BCHD_HOST` here.

jcramer reviewed (changes requested):
The way this new feature currently is setup to operate will require the admin to run the distribute function every 18 NFTs generated.  The NFT faucet can be done in a way where this burden to a token operator can be eliminated.

IMO we should distribute the Group tokens in a way just like the Token Type 1 tokens are distributed.  Then when user submits their address for an NFT you can do a Group NFT send to isolate qty 1 for burning, then send that qty 1 in a NFT GENESIS for burning.  The GENESIS NFT should output to the user's address.

zh commented:
This code require too many changes to the original code. Closed in favours of  [PR#3](https://github.com/simpleledger/slp-faucet/pull/3) code.
